docs/contributor/writing-effective-tests.md

```diff
@@ -0,0 +1,359 @@
+
+```typescript
+import { expect } from "chai";
+import { mockFn, mockObject, instance } from "../MockUtils";
+
+class TestClass {
+    private first: number;
+    private second: number;
+
+    constructor(first: number, second: number) {
+        this.first = first;
+        this.second = second;
+    }
+
+    sum(): number {
+        return this.first + this.second;
+    }
+}
+
+function sumOfTestClass(test: TestClass): number {
+    return test.sum();
+}
+
+
+test("can mock a class with private properties", () => {
+    const mockedClass = mockObject<TestClass>({
+        sum: mockFn();
+    });
+
+    mockedClass.sum.returns(42);
+    expect(sumOfTestClass(instance(mockedInterface))).to.equal(42);
+});
+```
+
+### Mocking VS Code events
+
+The `AsyncEventEmitter` captures components listening for a given event and fires the event emitter with the provided test data:
+
+```typescript
+import { expect } from "chai";
+import { mockObject, AsyncEventEmitter } from "../MockUtils";
+import * as vscode from "vscode";
+
+interface TestInterface {
+    onDidGetNumber: vscode.Event<number>;
+}
+
+test("example of mocking an asynchronous event within a mocked object", async () => {
+    // Create a new AsyncEventEmitter and hook it up to the mocked interface
+    const emitter = new AsyncEventEmitter<number>();
+    const mockedInterface = mockObject<TestInterface>({
+        onDidGetNumber: mockFn(s => s.callsFake(emitter.event));
+    });
+
+    // A simple example of an asynchronous event handler
+    const events: number[] = [];
+    mockedInterface.onDidGetNumber(async num => {
+        await new Promise<void>(resolve => {
+            setTimeout(resolve, 1);
+        });
+        events.push(num);
+    });
+
+    // Use the fire() method to trigger events
+    await emitter.fire(1);
+    await emitter.fire(2);
+    await emitter.fire(3);
+
+    // Make sure that the events were triggered correctly
+    expect(events).to.deep.equal([1, 2, 3]);
+});
+```
+
+### Mocking global modules
+
+Sometimes it is necessary to mock behavior that is provided by modules. A prime example of this is the VS Code API. In these cases you can use the global variants of the previously mentioned functions.
+
+These global mocking functions automatically handle `setup()` and `teardown()` through Mocha so that you don't have to do this yourself. This greatly simplifies writing tests. For example, you _could_ write a test that mocks a piece of the VS Code API like so:
+
+```typescript
+import { expect } from "chai";
+import { mockFn, mockObject, MockedObject } from "../MockUtils";
+import * as vscode from "vscode";
+
+suite("Mocking a Global Interface", async function () {
+    let mockedVSCodeWindow: MockedObject<typeof vscode["window"]>;
+    let originalWindowObject: typeof vscode["window"];
+
+    setup() {
```

Comment:
wouldn't this setup and teardown be hidden behind a utility function? Would be more prone to errors if developers had to do this themselves

Reply:
Yeah the continuation after this example shows how using `mockGlobalObject()` is easier. This was just a hypothetical example to show why `mockGlobalObject()` is easier to work with:

```markdown
However, you can remove the need for custom `setup()` and `teardown()` methods ...
```

Reply:
I missed the wording. I don't think we need to show the anti-pattern

Reply:
Makes sense. I replaced that code example with the mockGlobalObject() one pointing out how it must be called at the `suite()` level since it handles `setup()` and `teardown()` for you.